Add ramps 1.4.4 to AGCM4

GMagician opened the pull request:
Adding first shield to AGCM. I Hope to put my hands on it as soon as @MrAlvin will be ready and to finally test a real system

@MrAlvin if you have some time please validate pinout

~~N.B. current SAMD framework has an issue with usb serial and Marlin will "forget" some chars when send to host. Fix has been already merged into new adafruit framework, but PlatformIO has not been updated yet~~

It seems that now adafruit/platformio has switched from TINYUSB to USBCON then above statement is no more relevant

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -1,4 +1,4 @@
-/**
+﻿/**
```
Has the encoding been changed on this or a BOM added?

GMagician replied:
I think my editor did it (added BOM). To properly display Prusa I have to save it in utf8..I didn't revert code using git but only restoring content and this may be the result.

thinkyhead commented on Marlin/src/pins/samd/pins_RAMPS_144.h:
```diff
@@ -0,0 +1,547 @@
+    #define Y_SERIAL_TX_PIN  ?59
+  #endif
+  #ifndef Y_SERIAL_RX_PIN
+    #define Y_SERIAL_RX_PIN  ?64
```
I don't think we can leave these `?` characters in the file.

GMagician replied:
I know, this is why I asked to @MrAlvin to take a look at this. I don't have full board schematic and I need his help

MrAlvin commented:
Hi @GMagician,
good to see that you are still going strong. 

I am not quite sure how to follow up on this request. But I think you would like me to look at the pin assignments where there are question marks, correct?

MrAlvin commented:
I am not sure how to verify these three. 

For the RAMPS 1.4.4 shield, the software pins going to the TMC2130 drivers, are the same pins as the hardware pins. So I would think that they need to be:
MOSI  51
MISO  50
SCK   52
(as also indicated by this page: https://www.arduino.cc/en/reference/SPI )

Unless you choose to add Dupont wires to the top of the stepper drivers. And then it is possible to use any available pins. 


#if ENABLED(TMC_USE_SW_SPI)
  #ifndef TMC_SW_MOSI
    #define TMC_SW_MOSI    51 //?66
  #endif
  #ifndef TMC_SW_MISO
    #define TMC_SW_MISO    50 //?44
  #endif
  #ifndef TMC_SW_SCK
    #define TMC_SW_SCK     52 //?64
  #endif
#endif

MrAlvin commented:
About: 
  #ifndef X_SERIAL_TX_PIN
    #define X_SERIAL_TX_PIN  ?40
  #endif
  #ifndef X_SERIAL_RX_PIN
    #define X_SERIAL_RX_PIN  ?63
  #endif


Is it still necessary to define two pins for Rx/Tx  to the 220x PDN UART signal? 

I would suggest to wait with assignments, until it is possible to use only one MCU pin for 220x UART communication. 


A general note about which MCU pin to use for PDN UART with the 220x drivers.

As I understand, then It is not universally defined which pin is the PDN pin (UART pin) on the 220x step-stick boards/revisions of the boards. 

[Edit] 
Examples of how PDN is implemented on different StepStick designs is collected here: 
https://reprap.org/wiki/TMC2209
[/Edit]

But is seems that the newest 220x step-stick boards have an option to chose between some pins going to the shield, for the PDN signal, so I would suggest to use the pin that is otherwise used for CS on the TMC2130 drivers, as it is easy to connect that pin from the step-stick boards to a MCU pin, via the RAMPS 1.4.4 shield. 

[Edit] 
The CS pin is otherwise also known as the MS3 pin 
[/Edit]

This would result in the following definitions: 
X_PDN 47
Y_PDN 45
Z_PDN 32
E0_PDN 22
E1_PDN 43

For some MCU's it might however not be possible/feasable to use those pins as a universal one-wire Rx/Tx Uart software solution, as it might be necessary to be able to attach an interrupt routine to the MCU pin used. 

I am not sure if the AGCM4 board has interrupt abilities on these pins.

[Edit]
It seems like both the Due and the AGCM4 MCUs have an option to have an interrupt associated with any pin on the chip
[/Edit]

[Edit] 
Provisions are also now made on the shield, so it might be possible to use either pin D47 or pin D66 as a Single SoftwareSerial for up to four TMC2209 drivers.
[/Edit]

MrAlvin commented:
About  // LCDs and Controllers //

I only have access to the 
RepRapDiscount Smart Controller
and the 
RepRapDiscount FULL GRAPHIC Smart Controller
so they are the only LCDs that I have verified to work with the RAMPS 1.4.4 shield.  

But as AUX-1, AUX-2 AUX-3 and AUX-4 are the same as the original RAMPS 1.4 shield, the other LCD solutions/definitions should still work the same, and would typically use the same pin assignments as the pins_RAMPS_14.h files has defined them. 


Additionally I am however also preparing an LCD adapter board with a built in I2C MCP23017 chip, and  this LCD adapter is intended to be usable in combination with LCDs that use the EXP-1 and EXP-2 sockets, so I will also be giving the I2C LCD options some attention, at a later date. 

The purpose of this new LCD adapter board is, to free some pins on AUX-4, so it becomes possible to control more stepper drivers (or something else) from a RAMPS based setup.

MrAlvin commented:
@GMagician Did these comments clear up some of your questions? 

Also, I am right now soldering the shield that is intended to be shipped to you. Ok?

GMagician commented:
@MrAlvin thanks for your support, you are right about why I involved you.

About mosi: I'll define here default pins. Pin 50, 51, 52 are SPI bus and, on AGCM4, are the same H/W pins connected to "central connector". Adafruit has also named these pin with different numbering then depending on where ramps is connected to I would use proper numbers (I know this is just a pedantry).
About 220x UART: SAMD51 uses a 'sercom' module to handle all serial buses (I2C, SPI, UART) so to properly handle UART for 220x stepstick only its dedicated pin needs to be connected (otherwise a soft uart handling must be written). Now SPI is also used by LCD SD (don't know if also used by some LCD themselves) so I think that 220x stepsticks will require different pins to be used in UART mode. I may decide to not define "defaults" for them.
~~About your LCD: Better to share SD SPI for it (and a different CS pin) so that both AGCM4 and DUE may use H/W spi instead of soft one~~
Edit: I re-read your post. Your LCD works with i2c, not spi...ok

GMagician commented:
@MrAlvin 

> Also, I am right now soldering the shield that is intended to be shipped to you. Ok?

ok, please don't solder stepper, endstops, thermistores, dupont headers, I will replace them (where possible) with 5120 ones.

Edit: PM me to details about this

GMagician commented:
@MrAlvin please take also a look to other defined pins, just to be sure they are correct (please take care of different analog pins numbering AGCM4 has when they are used as digital)

GMagician committed:
Add ramps 1.4.4 to AGCM4

GMagician commented:
@MrAlvin I changed almost all pins to AGCM4, even soft UART (they are on aux2 and you said this connector is compatible with ramps).

Only doubt is SDSS and SPI (TMC spi). Where are they connected? To central connector, to 50-52 or free (solder/jumper needed)

GMagician committed:
Complete definition

MrAlvin commented:
@GMagician I have sent you a PM, and included schematics and PCB design files. 
And I see that you have already been very fast and have updated your Marlin files. Way to go :-)

thinkyhead committed:
Update pins_RAMPS_144.h

GMagician committed:
Add extruders 5-7 TX/RX pins

GMagician committed:
(c) 2020

GMagician committed:
Add some comments

GMagician committed:
Fix swapped extruders

Swapped E0 with E1 pins
Renamed E1 to Z2 since without hotend E1 is useless
Use stepsticks CS as soft serial TX/RX
Add correct onboard SD handling

GMagician commented:
Updates: 
 I tested my lcd and its SD card and they are working
 I tested onboard SD card and it's working
 I tested fans and they are working
 I tested hotend and bed they are working
 I tested thermistores and they are working (when other AGCM4 PR will be merged)
I tested eeprom and it's working
**Endstop are not working but I cant' find where issue is coming from..investigating**
Servos not tested yet
Stepper not tested

GMagician commented:
> Endstop are not working but I cant' find where issue is coming from..investigating

this seems a ramps hardware issue. @MrAlvin already found out how to solve it

GMagician commented:
@thinkyhead I don't have any bltouch 3.3V so I can't test servos part. It may not works since has been developped using original servo lib (at the time of developping original lib wasn't working at all on samd51), also stepper, I need to implement SoftwareSerial but I still don't have any tmc2208/2209 (I have to buy one to do tests), nor I have tmc2130 or any free stepstick (I have only drv8825 but are connected to my printer). No step motors (only on printer)

Everything else is now working (endstops issue was hardware cause)

GMagician commented:
@thinkyhead I think this can be merged now. Tested servo (connected bl-touch without input) all commands are working as expected.

thinkyhead commented:
Great, thanks for keeping it moving forward. I'll give it a once-over review now….
